### ISSUE TITLE
Potential to integrate with rclone mounts 

### ISSUE THREAD

**Issue by daniel-loader** [l3uddz/plex_autoscan#20]
Since rclone now has the ability to locally purge some of the cache on a remote mount, if you're using google drive for example to host your media, it may be helpful to those who have a seedbox and a cloud plex VPS.

In my own use case I have a seedbox and a plex vps, and I have a 10 min wait time for plex_autoscan to attempt to find the files but it doesn't always purge the cache in time so plex_autoscan eventually gives up and the files aren't added to plex. 

https://github.com/ncw/rclone/blob/master/docs/content/rc.md

**Comment by l3uddz** [comment]
So like an option to use the vfs/forget endpoint when it does not detect a file? Able to test that endpoint on a specific file as this occurs and let me know if that indeed causes just that files cache entry to be dropped then refreshed when autoscan checks if the file exists before it proceeds with a scan?

**Comment by daniel-loader** [comment]
I'll investigate tonight, but yeah essentially it would be a step prior to checking for the file. 
I'll try to trigger the event between one of the checks.

**Comment by l3uddz** [comment]
Ok cool, well if it works as expected I'll add it np

**Comment by daniel-loader** [comment]
Would you prefer to implement it as a calling a local rclone command to interface with it or curl-ing to the http listener? I'll check whichever suits best.

**Comment by l3uddz** [comment]
will probably send an HTTP request, so if you are able to test it with curl that would be ideal. 
i think the ideal one to check first would be vfs/forget as hopefully that works as expected (removes it from cache so further file requests on that location cause it to refresh, thus showing the file). 

Could you check if the file parameter can provide a full file path such as the one being checked for existence, if that works then great and will likely be a quick addition.

I guess the other would be the purge command, although that would be more tricky as autoscan has no concept of rclone remotes.

**Comment by daniel-loader** [comment]
Could it not be handled in a similar way to the existing path remapping for different path structures that exist between sonarr and the Plex located server? 

On my sonarr box it's /home/htpc/media and on Plex it's /home/plex/media and it works fine. Guess you'd need to take the path of the file that's relative to that root, remove the prefix and have a variable for the rclone remote name.

So in my case /home/plex/media/TV/Example/file s01e01.mkv would be $rclonevariable=/TV/Example/file s01e01.mkv

**Comment by l3uddz** [comment]
yep, would likely be the route taken. although

```rclone rc vfs/forget file=hello file2=goodbye dir=home/junk```

gives me the impression passing the mounted file location of the file that is being checked, should cause it to forget that location, and hopefully when autoscan performs the file check, will cause rclone to refresh that directory/file thus appearing if it exists and continue with the scan.

**Comment by daniel-loader** [comment]
Yeah that's essentially the behaviour I'm looking for, prior to --rc flag being implemented I was having to run a `killall -HUP rclone` signal to purge the cache, but sadly that took the data in addition to the directory tree cache. So if it's purging it and then the remote is polled for the change it should, work as expected on the second check.

**Comment by daniel-loader** [comment]
Okay few notes: You don't need to specify the remote or have any concept of remotes, the remote is determined by the port used to curl it - the mount command sets the port for the listener (default 5572).

Example:

```# curl -X POST 'http://localhost:5572/cache/expire/?remote=TV/Krypton/'
{
        "message": "cached directory cleared: dggdoar82340bl4k78kghr954c/puecl8v34git6993p00ku761eg/",
        "status": "ok"
}
```

So as long as you can pass the path relative component from the mapped root to curl with the right port then it should expire as expected.

**Comment by l3uddz** [comment]
Awesome thanks, able to post the full curl command used. and the full location of the file that is being scanned / checked exists?

**Comment by daniel-loader** [comment]
Sure I'm about to do some checks by manually running through a process of adding a new TV show, downloading/importing/uploading then interjecting the cache eviction commands while plex_autoscan does it multiple checks to find it. 

One thing to consider is having two purges, (assuming the plex_autoscan config is set up to do multiple retries.)

One after the first check to purge the destination directory `TV/Show/Season/Episode.mkv` and then on the penultimate check purge `TV/Show/Season/`. 

Rationale being that if it's a new season, purging the directory cache on the mount won't catch the changes as the parent will have the directory structure cached with no information about the new child.

**Comment by l3uddz** [comment]
thanks for the data, was helpful in writing a crude function up that does this (atleast in theory, will need actual testing with a live rclone instance if your up for it :P)

you can checkout the dev branch now, it will add a new option to your config:
```json
  "RCLONE_RC_CACHE_EXPIRE": {
    "ENABLED": false,
    "MOUNT_FOLDER": "/mnt/rclone",
    "RC_URL": "http://localhost:5572"
  },
```

you will need to enable it and change the mount folder for your setup, likely to ```/home/plex/media```

Please let me know how it goes, any issues etc as its untested against, so will probably need some adjustments.

**Comment by daniel-loader** [comment]
Done as requested, will report how it goes!
Does it handle escaping spaces and brackets? I couldn't work out how to do it via curl, it worked "as expected" when doing it via bash to `rclone rc` though.

**Comment by l3uddz** [comment]
The python requests module should correctly format it so that it posts correctly, although I'm not sure if it still requires escaping of special characters for cache to find it. 

If it does I can do that no problem, let me know of any iasues such as that occur during your tests.

**Comment by daniel-loader** [comment]
Okay so tested it for a few days, only thing that stands out is the fact the logs indicate an unexpected response when the response is ok.

```
Mar 28 02:32:28 plex python[15321]:  2018-03-28 02:32:28,175 -     INFO -  AUTOSCAN [140566874863360]: Client '185.225.208.27' scan request for series: '/home/htpc/media/TV/Shadowhunters/Shadowh
unters - S03E02 - The Powers That Be WEBDL-720p.mkv', event: 'Download'
Mar 28 02:32:28 plex python[15321]:  2018-03-28 02:32:28,178 -     INFO -      PLEX [140566811854592]: Scan request for '/home/plex/media/TV/Shadowhunters/Shadowhunters - S03E02 - The Powers Tha
t Be WEBDL-720p.mkv', sleeping for 630 seconds...
Mar 28 02:42:58 plex python[15321]:  2018-03-28 02:42:58,285 -     INFO -      PLEX [140566811854592]: File '/home/plex/media/TV/Shadowhunters/Shadowhunters - S03E02 - The Powers That Be WEBDL-7
20p.mkv' did not exist on check 1 of 20, checking again in 60 seconds.
Mar 28 02:43:58 plex python[15321]:  2018-03-28 02:43:58,290 -     INFO -     UTILS [140566811854592]: Sending rclone cache clear to http://localhost:5572/cache/expire for: 'TV/Shadowhunters'
Mar 28 02:43:58 plex python[15321]:  2018-03-28 02:43:58,432 -    ERROR -     UTILS [140566811854592]: Unexpected rclone cache clear response from http://localhost:5572/cache/expire while trying
 to clear 'TV/Shadowhunters': {
Mar 28 02:43:58 plex python[15321]:         "message": "cached directory cleared: dggdoar82340bl4k78kghr954c/95dv2vpodje1dhv4qvo6nhnnsk",
Mar 28 02:43:58 plex python[15321]:         "status": "ok"
Mar 28 02:43:58 plex python[15321]: }
Mar 28 02:44:16 plex python[15321]:  2018-03-28 02:44:16,613 -     INFO -      PLEX [140566811854592]: File '/home/plex/media/TV/Shadowhunters/Shadowhunters - S03E02 - The Powers That Be WEBDL-7
20p.mkv' exists on check 2 of 20.
Mar 28 02:44:16 plex python[15321]:  2018-03-28 02:44:16,614 -     INFO -      PLEX [140566811854592]: Scan request is now being processed
Mar 28 02:44:16 plex python[15321]:  2018-03-28 02:44:16,630 -     INFO -      PLEX [140566811854592]: No 'Plex Media Scanner' processes were found.
Mar 28 02:44:16 plex python[15321]:  2018-03-28 02:44:16,630 -     INFO -      PLEX [140566811854592]: Starting Plex Scanner
```

**Comment by l3uddz** [comment]
Good to hear. I have made another small change r.e. detection of successful cache clear responses.
If you could git pull the latest and try that out for me, would be great!

**Comment by daniel-loader** [comment]
Tried a movie (name changed)

```
Mar 28 13:59:23 plex python[5725]:  2018-03-28 13:59:23,576 -     INFO -  AUTOSCAN [140705964517120]: Client '185.225.208.27' scan request for movie: '/home/htpc/media/Movies/Movie (2018)/Movie (2018) WEBDL-720p.mkv', event: 'Download'
Mar 28 13:59:23 plex python[5725]:  2018-03-28 13:59:23,578 -     INFO -      PLEX [140705901508352]: Scan request for '/home/plex/media/Movies/Movie (2018)/Movie (2018) WEBDL-720p.mkv', sleeping for 630 seconds...
Mar 28 14:09:53 plex python[5725]:  2018-03-28 14:09:53,656 -     INFO -      PLEX [140705901508352]: File '/home/plex/media/Movies/Movie (2018)/Movie (2018) WEBDL-720p.mkv' did not exist on check 1 of 20, checking again in 60 seconds.
Mar 28 14:10:53 plex python[5725]:  2018-03-28 14:10:53,713 -     INFO -     UTILS [140705901508352]: Sending rclone cache clear for: 'Movies/Movie (2018)'
Mar 28 14:10:53 plex python[5725]:  2018-03-28 14:10:53,733 -     INFO -     UTILS [140705901508352]: Failed to clear rclone cache for 'Movies/Movie (2018)': remote control command failed: 5d8b8ektqup7hrju14dgpr5ehc/1nbf85qtc79scc25cukpsuc9pv5eibiit5qjrqnh47fbc4hpa3n0 doesn't exist in cache
Mar 28 14:10:53 plex python[5725]:  2018-03-28 14:10:53,733 -     INFO -     UTILS [140705901508352]: Sending rclone cache clear for: 'Movies'
Mar 28 14:10:53 plex python[5725]:  2018-03-28 14:10:53,839 -     INFO -     UTILS [140705901508352]: Successfully cleared rclone cache for 'Movies'
Mar 28 14:12:09 plex python[5725]:  2018-03-28 14:12:09,495 -     INFO -      PLEX [140705901508352]: File '/home/plex/media/Movies/Movie (2018)/Movie (2018) WEBDL-720p.mkv' exists on check 2 of 20.
Mar 28 14:12:09 plex python[5725]:  2018-03-28 14:12:09,496 -     INFO -      PLEX [140705901508352]: Scan request is now being processed
Mar 28 14:12:09 plex python[5725]:  2018-03-28 14:12:09,512 -     INFO -      PLEX [140705901508352]: No 'Plex Media Scanner' processes were found.
Mar 28 14:12:09 plex python[5725]:  2018-03-28 14:12:09,513 -     INFO -      PLEX [140705901508352]: Starting Plex Scanner
Mar 28 14:12:14 plex python[5725]:  2018-03-28 14:12:14,576 -     INFO -     UTILS [140705901508352]: GUI: Scanning Movie (2018)
Mar 28 14:12:20 plex python[5725]:  2018-03-28 14:12:20,675 -     INFO -     UTILS [140705901508352]: GUI: Matching 'Movie'
Mar 28 14:12:27 plex python[5725]:  2018-03-28 14:12:27,124 -     INFO -     UTILS [140705901508352]: GUI: Score for 'Movie' (2018) is 100
Mar 28 14:12:27 plex python[5725]:  2018-03-28 14:12:27,388 -     INFO -     UTILS [140705901508352]: GUI: Requesting metadata for 'Movie'
Mar 28 14:12:27 plex python[5725]:  2018-03-28 14:12:27,539 -     INFO -     UTILS [140705901508352]: GUI: Background media analysis on Movie
Mar 28 14:12:49 plex python[5725]:  2018-03-28 14:12:49,188 -     INFO -      PLEX [140705901508352]: Finished scan!
Mar 28 14:12:49 plex python[5725]:  2018-03-28 14:12:49,188 -     INFO -      PLEX [140705901508352]: Checking deleted item count in 10 seconds...
Mar 28 14:12:59 plex python[5725]:  2018-03-28 14:12:59,360 -     INFO -      PLEX [140705901508352]: Skipping emptying trash as there were no deleted items
```

Worked as expected, will keep an eye on it!

**Comment by l3uddz** [comment]
Glad to hear

**Comment by Krandor1** [comment]
Just installed this app for this very feature and working great and love the scan just the file/folder added vs having to re-scan my whole library.  Thanks for the app.

**Comment by daniel-loader** [comment]
Just a thought I had today when I was doing some tweaking with playback performance.

Executed:
`find /home/plex/media/Movies -type f -mmin -1440 -print -exec head -c 32M {} > /dev/null \;`

As expected it printed a list of all the files it could find under one day old and then used head to grab the cache block size I have set, managed to shave a few seconds off playback starting as the cache was warm for the first block of those files.

So it makes me wonder if a final config flag could be to "prewarm" the cache up, sure this is just a wishlist function but rclone integration was a week ago!

The sticking point would be whether you'd want to `head` a fixed size or a variable, though I suspect just opening the file would do if your mount options for the buffer + config option for cache chunk size are the same as it'll race to fill the buffer (and then the cache chunk after).  

Will do a little bit of testing. Figured it was worth keeping it here as it's under the umbrella of rclone integration but can open another issue if it's worth pursuing - honestly though it's just a fringe idea I had for warming up the cache on import.

**Comment by l3uddz** [comment]
Will think about how to add it. In the meantime there is already RUN_COMMAND_BEFORE_SCAN that might help, granted you wouldn't want it ran every time before a scan which that command does.

**Comment by l3uddz** [comment]
Also I would assume the actual Plex scan of the file, would trigger a small amount of the file to be read when it analyses the file?

**Comment by daniel-loader** [comment]
Yeah that's quite probable, seems unnecessary then unless you wanted something more complete like a full buffer of the file into the cache on import.

**Comment by l3uddz** [comment]
I could adjust that run command before scan config option so you can use a variable like ‰SCAN_FILE‰ then you could have it do as you just said, buffer a small bit before the scan

**Comment by l3uddz** [comment]
Which could be useful, especially when you have the analyze file Plex autoscan setting enabled
